### PR TITLE
fix: recompile artefacts from a different noir version

### LINKED ISSUE #3130: Serialization changes do not recompute the artifacts if it already exists on file
### Aim

This bug was caught by Martín Verzilli.

When a serialization change is made, nargo does not check if the json artifact with the old bytecode is not compatible.

This can lead to users running `nargo execute` and arriving the program panicking when trying to deserialize the old bytecode with the new serialization format.

### Expected Behavior

When a new noir version is made, regardless of whether it changed serialization, we should recompile the circuit. This also applies for nightly versions.

To get there, we need to tag each json file with an extra field "noir_version" and possibly "acvm_version". When we compile, we can check if the artifact is already there and then check the noir_version. If the noir_version has changed, then we recompile.

### Bug

Currently it does not recompile.

### To Reproduce

1.
2.
3.
4.

### Installation Method

None

### Nargo Version

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #3130 

## Summary\*

We do not use the cached artefacts if they were generated by a different noir version.
## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
Would this work for a locally dirty branch as well? It looks like it is based off the cargo package version so I'd assume not.

File: compiler/wasm/src/compile.rs [hunk 1]
```diff
@@ -17,6 +17,9 @@ use crate::errors::JsCompileError;
 
 const BACKEND_IDENTIFIER: &str = "acvm-backend-barretenberg";
 
+const NOIR_ARTIFACT_VERSION_STRING: &str =
```
Comment (kevaundray) [inline compiler/wasm/src/compile.rs:20]: We can probably define this in noirc_driver, then re-export it from there to Nargo and noir_wasm

File: ex/noir-trie-proofs
Comment (kevaundray) [inline ex/noir-trie-proofs:1]: this seems to be an extra file

Comment (guipublic) [reply ex/noir-trie-proofs:1]: indeed, I removed it.

**Comment by kevaundray** [comment]
Note that contracts were always recompiled, so this check only applies to programs -- we can open an issue about contracts since its out of scope for this PR

**Review by kevaundray — APPROVED** [review]

File: compiler/noirc_driver/src/lib.rs [hunk 2]
```diff
@@ -28,6 +28,15 @@ pub use program::CompiledProgram;
...
 const STD_CRATE_NAME: &str = "std";
 
+pub const GIT_COMMIT: &str = env!("GIT_COMMIT");
+pub const GIT_DIRTY: &str = env!("GIT_DIRTY");
+pub const NOIRC_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+/// Version string that gets placed in artifacts that Noir builds.
+/// Note: You can't directly use the value of a constant produced with env! inside a concat! macro.
+pub const NOIR_ARTIFACT_VERSION_STRING: &str =
+    concat!(env!("CARGO_PKG_VERSION"), "-", env!("GIT_COMMIT"));
```
Comment (kevaundray) [inline compiler/noirc_driver/src/lib.rs:38]:
```suggestion
/// Version string that gets placed in artifacts that Noir builds. This is semver compatible.
/// Note: You can't directly use the value of a constant produced with env! inside a concat! macro.
pub const NOIR_ARTIFACT_VERSION_STRING: &str =
    concat!(env!("CARGO_PKG_VERSION"), "+", env!("GIT_COMMIT"));
```

Comment (kevaundray) [reply compiler/noirc_driver/src/lib.rs:38]: The + string appended after the plus is seen as build metadata

**Pull request merged**